### PR TITLE
Backup etcd files locally

### PR DESCRIPTION
**Pull request by lewismarshall**

### CODE REVIEW

**Comment by vaijab** [comment]
Before a cluster upgrade, we want to be able to take a backup of etcd. Maybe exec into container, touch a file and then etcd backup is taken would be good enough?

**Comment by lewismarshall** [comment]
@vaijab added as an issue #2 . This PR is just first bit of work with docs. I'll definitely add the feature though...

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+FROM alpine:3.5
+ENV ETCD_VER v2.3.7
+RUN apk add --update bash curl jq
```
Comment (vaijab) [inline Dockerfile]: Add `--no-cache`

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+RUN curl -L  https://github.com/coreos/etcd/releases/download/${ETCD_VER}/etcd-${ETCD_VER}-linux-amd64.tar.gz -o /tmp/etcd-${ETCD_VER}-linux-amd64.tar.gz && \
+    tar -xzvf /tmp/etcd-${ETCD_VER}-linux-amd64.tar.gz -C /usr/local/bin && \
+    rm /tmp/etcd-${ETCD_VER}-linux-amd64.tar.gz && \
+    ln -s /usr/local/bin/etcd-${ETCD_VER}-linux-amd64/etcdctl /usr/local/bin/ && \
+    ln -s /usr/local/bin/etcd-${ETCD_VER}-linux-amd64/etcd /usr/local/bin/
+
+RUN apk -Uuv add groff less python py-pip && \
+    pip install awscli && \
+    apk --purge -v del py-pip && \
+    rm /var/cache/apk/*
```
Comment (vaijab) [inline Dockerfile]: `--no-cache` does that.

File: bin/backup [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+function error_exit() {
+  echo "${1}"
+  exit 1
+}
+
+function gettime() {
+  date +%H%M
+}
+
+# Will clear down any previous backup's locally (If pushing files elsewhere)
```
Comment (vaijab) [inline bin/backup]: backups, not backup's

File: bin/backup [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    fi
+  done
+  for backuptime in ${NODE_BACKUP_TIMES} ; do
+    if istime ${backuptime} ; then
+      echo "Time:$(gettime)"
+      nodebackup
+      sleep 60 # Ensure we skip past the checked time...
+    fi
+  done
+  sleep 15 # four attempts a minute to match a given time
```
Comment (vaijab) [inline bin/backup]: That's prone to fail. Your `istime()` matches on exact times.

File: kube/etcd-backup-ds.yaml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      labels:
+        name: etcd-backup
+    spec:
+      hostNetwork: true
+      containers:
+        - name: etcd-backup
+          image: quay.io/ukhomeofficedigital/etcd-backup:latest
+          resources:
+            limits:
+              memory: "50Mi"
```
Comment (vaijab) [inline kube/etcd-backup-ds.yaml]: I'd say, give it a bit more, a hundred at least.

**Review by vaijab — CHANGES_REQUESTED** [review]

**Comment by gambol99** [comment]
how come you don't just use cron as the entrypoint rather than placing the scheduling in bash ...

File: bin/backup (hunk 4 above)
Comment (lewismarshall) [reply bin/backup]: I'm only checking minutes here and only sleep for 60 when I need to guarantee not running twice - will add a few comments as discussed to.

File: bin/backup (hunk 3 above)
Comment (lewismarshall) [reply bin/backup]: Will push squashed update

**Comment by lewismarshall** [comment]
@gambol99 - I guess I didn't want to deal with background and systemd dependencies. One for next time...

**Comment by lewismarshall** [comment]
@vaijab All changes in.

**Review by vaijab — APPROVED** [review]
Neat!

**Pull request merged**